spock-core/src/main/java/org/spockframework/runtime/model/IterationInfo.java

```diff
@@ -65,6 +92,19 @@ public String getName() {
     return dataValues;
   }
 
+  /**
+   * Return an unmodifiable map of data variable names to values with the same iteration
+   * order as {@link FeatureInfo#getDataVariables} and {@link #getDataValues()}.
+   *
+   * <p>If {@code getDataValues()} returns {@code null}, this method does too which is for example
```

Comment:
or it was a simple feature the also have IterationInfo

Reply:
They do, but with an empty `dataValues` array and thus an empty `dataVariables` map.
Or did I miss something?
```java
private static final Object[] EMPTY_ARGS = new Object[0];
...
private SpockNode describeSimpleFeature(UniqueId parentId, FeatureInfo feature) {
    IterationInfo iterationInfo = new IterationInfo(feature, 0, EMPTY_ARGS, 1);
```

Reply:
no, my bad

I'm still not 100% convinced that we need to duplicate the information though. What happens if an extension modifies `dataValues`, the data in the map will remain unchanged.

Reply:
Hm, is that a valid use-case?
Shouldn't extensions just add things like providing objects needed for test like mocks or similar?
Is an extension expected to modify the data variables that I defined inside my tests?
My intention was to provide an easily accessible map data variable names to values, as this was the second time I had to write such mapping (the other PR with data variables in precondition context).
I thought the data variable values should not be modified by an extension and thus calculated it in the constructor.
But if that is a valid and supported use-case, I'll change the method to build the map every time the method is called.

Reply:
Btw. this is conceptually problematic anyway. The array is taken from `IterationInfo` and directly given to `MethodInvocation`. This is what extensions operate on. If an extension modifies the existing array, this change is reflected in `IterationInfo`. But as soon as one extension needs to add an argument, it is supposed to copy the array and set the new one in the invocation. Now the bond is broken and further extensions do not influence the `IterationInfo`.

Reply:
See also this comment: https://github.com/spockframework/spock/issues/652#issuecomment-249808491
So maybe it would be be better to have the method in question here calculate new every time in case an extesion modified the data values array on `IterationInfo`, but at the time it is given to `IMethodInvocation` intentionally break the bond by creating a new array of the correct length for the method and copying over the values, then extensions can also simpler set additional arguments.

Reply:
I changed it to rebuild the map now, but I still think the bond to the arguments array in the invocation object should be broken early, but that has nothing to do with this PR.

Reply:
hm, at least I would initialize it lazy, many will use an explicit unroll pattern and then this is wasted effort. This should also be fine to do it lock-free as the computation is repeatable and multi-threaded access should be close to 0.

And the null case shouldn't really reach anyone, maybe it would be better if we checked `context.getErrorInfoCollector().hasErrors()` directly after `nextArgs(context, iterators)` in `PlatformParameterizedSpecRunner` to avoid passing null here.

Reply:
I'm not sure about that, at least it should always deliver a result now that is as correct as the data value array.

Reply:
Actually `hasErrors` is checked right after `nextArgs` was called and the `IterationInfo` and `IterationNode` were instantiated.

Reply:
Is this resolved now, that I build the map on the fly?

Reply:
ping :-)

Reply:
> Actually hasErrors is checked right after nextArgs was called and the IterationInfo and IterationNode were instantiated.

I know that it is checked directly after, I only said this so that we could add the Invariant, that it is never null and don't have to add code/docs to handle it.

> I would initialize it lazy, many will use an explicit unroll pattern and then this is wasted effort. This should also be fine to do it lock-free as the computation is repeatable and multi-threaded access should be close to 0.

Reply:
> Actually hasErrors is checked right after nextArgs was called and the IterationInfo and IterationNode were instantiated.

Ok, then I misread your answer. I will remove the null-check and the sentence from the JavaDoc.

> I would initialize it lazy, many will use an explicit unroll pattern and then this is wasted effort. This should also be fine to do it lock-free as the computation is repeatable and multi-threaded access should be close to 0.

Well, the effort is not wasted anymore, as it is not initialized lazy, but on-the-fly.
The objection that extensions could manipulate the data variables was valid, so it is better to rebuild the map when requested to be up-to-date at least to the time the method was called.
I could alternatively build a new `Map` implementation that delegates to the names list and the values array, then it actually is always up to date and does not need extra time to build the map, but I don't know how complex it is to provide a `Map` implementation. :-)

Reply:
Ok, I now made a custom `Map` implementation that is backed by the actual values array and made the instantiation lazy and one-time.
I also removed the `null` check and the part of the JavaDoc.
Now the method should hopefully be fine. :-)